# **Numpy**

In [2]:
import numpy as np

Essa lista contém questões práticas incluindo os seguintes tópicos com numpy.

* Criação de array e seus atributos, intervalos numéricos em numPy
* Slicing e indexação de NumPy Array.
* Manipulação, pesquisa, classificação e divisão de arrays.
* Funções matemáticas com arrays e broadcasting arrays NumPy. 

## **Crie um array inteiro 4x2 e mostre seus atributos**

Os elementos deven ser um tipo de int16 sem sinal. E imprima os seguintes atributos: formato do array; dimensões; comprimento de cada elemento do array.

In [ ]:
#########################################
#         Escreva o código aqui - Soluções Abaixo
#########################################

In [ ]:
'''SOLUÇÃO 1'''

# Partindo de um array
array1 = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
print(f'Array formado: \n \n {array1}')
print(f'\n Formato do array: {array1.shape}')
print(f'\n Dimensões do array: {array1.ndim}')
print(f'\n Número total de elementos no array: {array1.size}')
print(f'\n Comprimento de cada elemento do array: {len(array1[0])}')

Array formado: 
 
 [[1 2]
 [3 4]
 [5 6]
 [7 8]]

 Formato do array: (4, 2)

 Dimensões do array: 2

 Número total de elementos no array: 8

 Comprimento de cada elemento do array: 2


In [ ]:
'''SOLUÇÃO 2'''
# Usando o random de números inteiros

from numpy import random
array2 = random.randint(100, size=(4, 2))

print(array2)

print(f'\n Formato do array: {array2.shape}')
print(f'\n Dimensões do array: {array2.ndim}')
print(f'\n Número total de elementos no array: {array2.size}')
print(f'\n Comprimento de cada elemento do array: {len(array2[0])}')

[[64 79]
 [88 73]
 [17 93]
 [60  0]]

 Formato do array: (4, 2)

 Dimensões do array: 2

 Número total de elementos no array: 8

 Comprimento de cada elemento do array: 2


##### Resultado esperado:
<pre>
Mostrando Array
[[64392 31655]
 [32579     0]
 [49248   462]
 [    0     0]]

Mostrando Attributos:

Formato do array:  (4, 2)
Número de dimensões:  2
Comprimento de cada elemento do array:  2
</pre>

## **Crie uma array de inteiros 5X2 com um intervalo entre 100 e 200, de modo que a diferença entre cada elemento seja 10**

In [ ]:
array3 = np.arange(100, 200, 10).reshape((5, 2))
print(array3)

[[100 110]
 [120 130]
 [140 150]
 [160 170]
 [180 190]]


##### Resultado esperado:
<pre>
[[100 110]
 [120 130]
 [140 150]
 [160 170]
 [180 190]]
</pre>

## **Baseado no array fornecido. Retorne um array de itens com a terceira coluna de todas as linhas**

In [ ]:
array_00 = np.array([[11 ,22, 33], [44, 55, 66], [77, 88, 99]])
array_00

array([[11, 22, 33],
       [44, 55, 66],
       [77, 88, 99]])

In [ ]:
'''SOLUÇÃO 1'''

array_00_1 = []
for i in range(len(array_00)):

  array_00_1.append(array_00[i][-1])
array_00_1  

[33, 66, 99]

In [ ]:
'''SOLUÇÃO 2'''

array_00_2 = map(lambda x: array_00[x][-1], range(len(array_00)))
print(list(array_00_2))

[33, 66, 99]


In [ ]:
'''SOLUÇÃO 3'''

array_00_3 = [val for rows in array_00 for val in rows if val == rows[-1]]
array_00_3

[33, 66, 99]

In [ ]:
'''SOLUÇÃO 4'''
array_00_4 = array_00[:,-1]
print(array_00_4)

[33 66 99]


##### Resultado esperado:
<pre>
Mostrando array de entrada
[[11 22 33]
 [44 55 66]
 [77 88 99]]

Novo array com a terceira coluna de cada linha
[33 66 99]
</pre>

## **Retorne um array com os valores presentes nas linhas ímpares com colunas pares baseado no seguinte array**

In [3]:
array_01 = np.array([[3 ,6, 9, 12], [15 ,18, 21, 24], 
[27 ,30, 33, 36], [39 ,42, 45, 48], [51 ,54, 57, 60]])
array_01

array([[ 3,  6,  9, 12],
       [15, 18, 21, 24],
       [27, 30, 33, 36],
       [39, 42, 45, 48],
       [51, 54, 57, 60]])

In [ ]:
array_01.shape

(5, 4)

In [ ]:
#########################################
#         Escreva o código aqui - Soluções Abaixo
#########################################

In [ ]:
'''SOLUÇÃO 1'''

array_01_1 = []
for i in range(len(array_01)):
  if i % 2 == 0:
    array_01_2 = []
    for j in range(len(array_01[0])):
      if j % 2 != 0:
        array_01_2.append(array_01[i][j])
    array_01_1.append(array_01_2)        
array_01_1  

[[6, 12], [30, 36], [54, 60]]

In [ ]:
'''SOLUÇÃO 2'''

array_01_3 = []
for i in range(0, 5, 2):
  v = array_01[[i],[1, 3]]
  array_01_3.append(v)

array_01_3  

[array([ 6, 12]), array([30, 36]), array([54, 60])]

In [36]:
'''SOLUÇÃO 3'''

array_01_4 = array_01[::2,1::2]
print(array_01_4)

[[ 6 12]
 [30 36]
 [54 60]]


##### Resultado esperado:
<pre>
Mostrando array de entrada
[[ 3  6  9 12]
 [15 18 21 24]
 [27 30 33 36]
 [39 42 45 48]
 [51 54 57 60]]

Array resultante (linha ímpar x coluna par)
[[ 6 12]
 [30 36]
 [54 60]]
</pre>

## **Crie um array de resultado somando os seguintes arrays. Em seguida, modifique o array resultante calculando o quadrado de cada elemento**

In [25]:
array_02 = np.array([[5, 6, 9], [21 ,18, 27]])
array_03 = np.array([[15 ,33, 24], [4 ,7, 1]])

In [ ]:
#########################################
#         Escreva o código aqui - Soluções Abaixo
#########################################

In [28]:
# SOMA ENTRE ARRAYS
array_soma = array_02 + array_03
print(array_soma)

[[20 39 33]
 [25 25 28]]


In [29]:
array_sum = np.sum([array_02, array_03], axis=0)
print(array_sum)

[[20 39 33]
 [25 25 28]]


In [ ]:
# QUADRADO DOS ELEMENTOS
array_quadrado = array_soma**2
print(array_quadrado)

[[ 400 1521 1089]
 [ 625  625  784]]


In [30]:
arrar_sqrt = np.square(array_sum)
print(arrar_sqrt)

[[ 400 1521 1089]
 [ 625  625  784]]


##### Resultado esperado:
<pre>
A soma dos arrays é: 
[[20 39 33]
 [25 25 28]]

Resultado depois de calcular o quadrado de cada número:
[[ 400 1521 1089]
 [ 625  625  784]]
</pre>

## **Faça a soma, subtração, multiplicação e divisão dos arrays**

In [44]:
array_04 = np.array([1, 2, 3])
array_05 = np.array([[1, 2, 3], [4, 5, 6]])

In [ ]:
#########################################
#         Escreva o código aqui - Solução abaixo
#########################################

In [ ]:
# SOMA DOS ARRAYS
'''SOLUÇÃO 1'''
array_soma = array_04 + array_05
print(array_soma)

[[2 4 6]
 [5 7 9]]


In [57]:
array_04_new = np.array([array_04, array_04])
array_04_new

array([[1, 2, 3],
       [1, 2, 3]])

In [49]:
'''SOLUÇÃO 2'''
array_sum = np.sum([array_04_new,array_05], axis=0)
print(array_sum)

[[2 4 6]
 [5 7 9]]


In [55]:
'''SOLUÇÃO 3'''
array_sum_2 = np.sum([[array_04, array_04], array_05], axis=0)
print(array_sum_2)

[[2 4 6]
 [5 7 9]]


In [54]:
'''SOLUÇÃO 4'''
array_sum_3 = np.sum([np.array([array_04, array_04]), array_05], axis=0)
print(array_sum_3)

[[2 4 6]
 [5 7 9]]


In [ ]:
# SUBTRAÇÃO DOS ARRAYS
'''SOLUÇÃO 1'''
array_diff = array_04 - array_05
print(array_diff)

[[ 0  0  0]
 [-3 -3 -3]]


In [62]:
'''SOLUÇÃO 2'''
array_diff_1= np.subtract(array_04_new, array_05)
print(array_diff_1)

[[ 0  0  0]
 [-3 -3 -3]]


In [ ]:
# MULTIPLICÃO DOS ARRAYS
'''SOLUÇÃO 1'''
array_mult = array_04 * array_05
print(array_mult)

[[ 1  4  9]
 [ 4 10 18]]


In [63]:
'''SOLUÇÃO 2'''
array_mult_1 = np.multiply(array_04_new, array_05)
print(array_mult_1)

[[ 1  4  9]
 [ 4 10 18]]


In [ ]:
# DIVISÃO DOS ARRAYS
'''SOLUÇÃO 1'''
array_div = array_04 / array_05
print(array_div)

[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]


In [64]:
'''SOLUÇÃO 2'''
array_div_1 = np.divide(array_04_new, array_05)
print(array_div_1)

[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]


##### Resultado esperado:
<pre>
Soma: 
[[2 4 6]
 [5 7 9]]
Subtração: 
[[ 0  0  0]
 [-3 -3 -3]]
Multiplicação: 
[[ 1  4  9]
 [ 4 10 18]]
divisão: 
[[1.   1.   1.  ]
 [0.25 0.4  0.5 ]]
</pre>

## **Divida o array em quatro sub-arrays de tamanhos iguais**

Crie um array de inteiros 8X3 com um intervalo entre 10 e 34 de modo que a diferença entre cada elemento seja 1 e, em seguida, faça a subdivisão. 

In [ ]:
#########################################
#         Escreva o código aqui - Solução abaixo
#########################################

In [ ]:
# Criando o array
array_completo = np.arange(10, 34, 1).reshape((8, 3))
print(array_completo)

[[10 11 12]
 [13 14 15]
 [16 17 18]
 [19 20 21]
 [22 23 24]
 [25 26 27]
 [28 29 30]
 [31 32 33]]


In [ ]:
# Usando o split (só consegue fazer a divisão das matriz se esta divisão for exata):
matriz_split = np.split(array_completo, 4)
print(matriz_split)

[array([[10, 11, 12],
       [13, 14, 15]]), array([[16, 17, 18],
       [19, 20, 21]]), array([[22, 23, 24],
       [25, 26, 27]]), array([[28, 29, 30],
       [31, 32, 33]])]


In [ ]:
# Usando o array_split (consegue fazer a divisão da matriz mesmo que esta não seja exata):

matriz_array_split = np.array_split(array_completo, 4)
print(matriz_array_split)

[array([[10, 11, 12],
       [13, 14, 15]]), array([[16, 17, 18],
       [19, 20, 21]]), array([[22, 23, 24],
       [25, 26, 27]]), array([[28, 29, 30],
       [31, 32, 33]])]


##### Resultado esperado:
<pre>
Criando o array de 8X3 usando numpy.arange
[[10 11 12]
 [13 14 15]
 [16 17 18]
 [19 20 21]
 [22 23 24]
 [25 26 27]
 [28 29 30]
 [31 32 33]]

Dividindo o array em 4 sub-arrays

[array([[10, 11, 12],[13, 14, 15]]), 
array([[16, 17, 18],[19, 20, 21]]), 
array([[22, 23, 24],[25, 26, 27]]), 
array([[28, 29, 30],[31, 32, 33]])]
</pre>

## **Ordenar o seguinte array da seguinte forma:**

* Oredene pela segunda linha
* Ordene pela terceira coluna

In [ ]:
array_06 = np.array([[34,43,73],[82,22,12],[53,94,66]])

In [ ]:
print(array_06)

[[34 43 73]
 [82 22 12]
 [53 94 66]]


In [ ]:
#########################################
#         Escreva o código aqui - Soluções abaixo
#########################################

### Ordenando pela segunda linha

In [ ]:
sort_row = array_06 [ :, array_06[1].argsort()]
print(sort_row)

[[73 43 34]
 [12 22 82]
 [66 94 53]]


### Ordenando pela terceira coluna

In [ ]:
'''SOLUÇÃO 1'''
sort_column = array_06[array_06[:, 2].argsort()]
print(sort_column)

[[82 22 12]
 [53 94 66]
 [34 43 73]]


In [ ]:
'''SOLUÇÃO 2'''
sort_1 = np.sort(array_06.view('i8,i8,i8'), order=['f2'], axis=0)
print(sort_1)

[[(82, 22, 12)]
 [(53, 94, 66)]
 [(34, 43, 73)]]


##### Resultado esperado:
<pre>
Mostrando o array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Ordenando o array pela segunda linha
[[73 43 34]
 [12 22 82]
 [66 94 53]]

Ordenando o array pela terceira coluna
[[82 22 12]
 [34 43 73]
 [53 94 66]]
</pre>

## **Imprima o máx. do eixo 0 e mín. Do eixo 1 do seguinte array**

In [ ]:
array_07 = np.array([[34,43,73],[82,22,12],[53,94,66]])
array_07

array([[34, 43, 73],
       [82, 22, 12],
       [53, 94, 66]])

In [ ]:
#########################################
#         Escreva o código aqui - Soluções abaixo
#########################################

In [ ]:
# Máximo no eixo 0
np.amax(array_07, axis=0) 

array([82, 94, 73])

In [ ]:
# Mínimo no eixo 1
np.amin(array_07, axis=1) 

array([34, 12, 53])

##### Resultado esperado:
<pre>
Mostrando array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Mostando amin do eixo 1
[34 12 53]

Mostrando amax do eixo 0
[82 94 73]
</pre>

## **Exclua a segunda coluna de um determinado array e insira a nova coluna seguinte em seu lugar**

In [ ]:
array_08    = np.array([[34,43,73],[82,22,12],[53,94,66]]) 
nova_coluna = np.array([[10,10,10]])

In [ ]:
array_08

array([[34, 43, 73],
       [82, 22, 12],
       [53, 94, 66]])

In [ ]:
#########################################
#         Escreva o código aqui - Solução abaixo
#########################################

In [ ]:
# Excluindo a segunda coluna:
array_08_del = np.delete(array_08, 1, 1)
array_08_del

array([[34, 73],
       [82, 12],
       [53, 66]])

In [ ]:
# Inserindo a nova coluna no lugar
array_08_new = np.insert(array_08_del, 1, nova_coluna, axis=1)
array_08_new

array([[34, 10, 73],
       [82, 10, 12],
       [53, 10, 66]])

##### Resultado esperado:
<pre>
Mostrando array original
[[34 43 73]
 [82 22 12]
 [53 94 66]]

Array depois de deletar a segunda coluna
[[34 73]
 [82 12]
 [53 66]]

Array depois de inserir a nova coluna no lugar da coluna excluída
[[34 10 73]
 [82 10 12]
 [53 10 66]]
</pre>